File: web/src/utils/itineraryUtils.ts
import { ItineraryItem, DayPeriod, TimeType } from "../data/models";

export function getEffectivePeriod(item: ItineraryItem): DayPeriod {
  if (item.timeType === TimeType.PERIOD && item.period) {
    return item.period;
  }

  const time = item.startTime;
  if (!time) return DayPeriod.MORNING; // Fallback

  const parts = time.split(":");
  const hour = parseInt(parts[0], 10) || 0;

  if (hour >= 5 && hour <= 10) return DayPeriod.MORNING;
  if (hour >= 11 && hour <= 15) return DayPeriod.NOON;
  if (hour >= 16 && hour <= 18) return DayPeriod.AFTERNOON;
  if (hour >= 19 && hour <= 21) return DayPeriod.EVENING;
  return DayPeriod.NIGHT;
}

export function getTimeHeaderLabel(item: ItineraryItem): string | null {
  if (item.timeType === TimeType.EXACT) return item.startTime || null;
  if (item.timeType === TimeType.RANGE) {
    if (item.startTime && item.endTime) {
      return `${item.startTime} - ${item.endTime}`;
    }
    return item.startTime || null;
  }
  return null;
}

export interface TimeGroup {
  label: string | null;
  items: ItineraryItem[];
}

export interface PeriodGroup {
  period: DayPeriod;
  timeGroups: TimeGroup[];
}

export function groupByHierarchy(items: ItineraryItem[]): PeriodGroup[] {
  const periods = [
    DayPeriod.MORNING,
    DayPeriod.NOON,
    DayPeriod.AFTERNOON,
    DayPeriod.EVENING,
    DayPeriod.NIGHT,
  ];

  return periods
    .map((p) => {
      const itemsInPeriod = items.filter((it) => getEffectivePeriod(it) === p);
      if (itemsInPeriod.length === 0) return null;

      const timeGroups: TimeGroup[] = [];
      let currentLabel: string | null = null;
      let currentItems: ItineraryItem[] = [];

      itemsInPeriod.forEach((item) => {
        const label = getTimeHeaderLabel(item);
        if (label === currentLabel && label !== null) {
          currentItems.push(item);
        } else {
          if (currentItems.length > 0) {
            timeGroups.push({ label: currentLabel, items: currentItems });
          }
          currentLabel = label;
          currentItems = [item];
        }
      });

      if (currentItems.length > 0) {
        timeGroups.push({ label: currentLabel, items: currentItems });
      }

      return { period: p, timeGroups };
    })
    .filter((g): g is PeriodGroup => g !== null);
}

File: app/src/main/java/com/tripro/app/util/ItineraryUtils.kt
package com.tripro.app.util

import com.tripro.app.data.model.DayPeriod
import com.tripro.app.data.model.ItineraryItem
import com.tripro.app.data.model.TimeType

object ItineraryUtils {

    fun getEffectivePeriod(item: ItineraryItem): DayPeriod {
        if (item.timeType == TimeType.PERIOD && item.period != null) {
            return item.period
        }
        
        val time = item.startTime ?: return DayPeriod.MORNING // Fallback
        val parts = time.split(":")
        val hour = parts.getOrNull(0)?.toIntOrNull() ?: 0
        
        return when (hour) {
            in 5..10 -> DayPeriod.MORNING
            in 11..15 -> DayPeriod.NOON
            in 16..18 -> DayPeriod.AFTERNOON
            in 19..21 -> DayPeriod.EVENING
            else -> DayPeriod.NIGHT
        }
    }

    fun getTimeHeaderLabel(item: ItineraryItem): String? {
        return when (item.timeType) {
            TimeType.EXACT -> item.startTime
            TimeType.RANGE -> {
                if (item.startTime != null && item.endTime != null) {
                    "${item.startTime} - ${item.endTime}"
                } else {
                    item.startTime
                }
            }
            TimeType.PERIOD -> null
        }
    }
}

data class TimeGroup(
    val label: String?,
    val items: List<ItineraryItem>
)

data class PeriodGroup(
    val period: DayPeriod,
    val timeGroups: List<TimeGroup>
)

fun List<ItineraryItem>.groupByHierarchy(): List<PeriodGroup> {
    return DayPeriod.entries.mapNotNull { p ->
        val itemsInPeriod = this.filter { ItineraryUtils.getEffectivePeriod(it) == p }
        if (itemsInPeriod.isEmpty()) return@mapNotNull null
        
        // Secondary grouping by time label
        val timeGroups = mutableListOf<TimeGroup>()
        var currentLabel: String? = null
        var currentItems = mutableListOf<ItineraryItem>()
        
        itemsInPeriod.forEach { item ->
            val label = ItineraryUtils.getTimeHeaderLabel(item)
            if (label == currentLabel && label != null) {
                currentItems.add(item)
            } else {
                if (currentItems.isNotEmpty()) {
                    timeGroups.add(TimeGroup(currentLabel, currentItems))
                }
                currentLabel = label
                currentItems = mutableListOf(item)
            }
        }
        if (currentItems.isNotEmpty()) {
            timeGroups.add(TimeGroup(currentLabel, currentItems))
        }
        
        PeriodGroup(p, timeGroups)
    }
}
